lib-ee/emqx_ee_connector/src/emqx_ee_connector_cassa.erl

```diff
@@ -0,0 +1,412 @@
+    on_stop/2,
+    on_query/3,
+    %% TODO: now_supported_now
+    %%on_batch_query/3,
+    on_get_status/2
+]).
+
+%% callbacks of ecpool
+-export([
+    connect/1,
+    prepare_sql_to_conn/2
+]).
+
+%% callbacks for query executing
+-export([query/3, prepared_query/3]).
+
+-export([do_get_status/1]).
+
+-type prepares() :: #{atom() => binary()}.
+-type params_tokens() :: #{atom() => list()}.
+
+-type state() ::
+    #{
+        poolname := atom(),
+        prepare_sql := prepares(),
+        params_tokens := params_tokens(),
+        %% returned by ecql:prepare/2
+        prepare_statement := binary()
+    }.
+
+-define(DEFAULT_SERVER_OPTION, #{default_port => ?CASSANDRA_DEFAULT_PORT}).
+
+%%--------------------------------------------------------------------
+%% schema
+
+roots() ->
+    [{config, #{type => hoconsc:ref(?MODULE, config)}}].
+
+fields(config) ->
+    cassandra_db_fields() ++
+        emqx_connector_schema_lib:ssl_fields() ++
+        emqx_connector_schema_lib:prepare_statement_fields().
+
+cassandra_db_fields() ->
+    [
+        {servers, servers()},
+        {keyspace, fun keyspace/1},
+        {pool_size, fun emqx_connector_schema_lib:pool_size/1},
+        {username, fun emqx_connector_schema_lib:username/1},
+        {password, fun emqx_connector_schema_lib:password/1},
+        {auto_reconnect, fun emqx_connector_schema_lib:auto_reconnect/1}
+    ].
+
+servers() ->
+    Meta = #{desc => ?DESC("servers")},
+    emqx_schema:servers_sc(Meta, ?DEFAULT_SERVER_OPTION).
+
+keyspace(type) -> binary();
+keyspace(desc) -> ?DESC("keyspace");
+keyspace(required) -> true;
+keyspace(_) -> undefined.
+
+%%--------------------------------------------------------------------
+%% callbacks for emqx_resource
+
+callback_mode() -> always_sync.
+
+-spec on_start(binary(), hoconsc:config()) -> {ok, state()} | {error, _}.
+on_start(
+    InstId,
+    #{
+        servers := Servers,
+        keyspace := Keyspace,
+        username := Username,
+        pool_size := PoolSize,
+        ssl := SSL
+    } = Config
+) ->
+    {ok, _} = application:ensure_all_started(ecpool),
+    {ok, _} = application:ensure_all_started(ecql),
+
+    ?SLOG(info, #{
+        msg => "starting_cassandra_connector",
+        connector => InstId,
+        config => emqx_misc:redact(Config)
+    }),
+
+    Options = [
+        {nodes, emqx_schema:parse_servers(Servers, ?DEFAULT_SERVER_OPTION)},
+        {username, Username},
+        {password, emqx_secret:wrap(maps:get(password, Config, ""))},
+        {keyspace, Keyspace},
+        {auto_reconnect, ?AUTO_RECONNECT_INTERVAL},
+        {pool_size, PoolSize}
+    ],
+
+    SslOpts =
+        case maps:get(enable, SSL) of
+            true ->
+                [
+                    %% note: type defined at ecql:option/0
+                    {ssl, emqx_tls_lib:to_client_opts(SSL)}
+                ];
+            false ->
+                []
+        end,
+
+    PoolName = emqx_plugin_libs_pool:pool_name(InstId),
```

Comment:
Note: `InstID` changes each time a bridge is (re)created, so this may lead to an atom leak.  It would be safer to set the pool name to the bridge ID (like `cassandra:some_name`), because that's stable between restarts.

Also, it's a different ID when probing as well: https://github.com/emqx/emqx/pull/10143#discussion_r1144727665